sa/sa.go

```diff
@@ -915,3 +915,188 @@ func (ssa *SQLStorageAuthority) Health(ctx context.Context) error {
 	}
 	return nil
 }
+
+// LeaseCRLShard marks a single crlShards row as leased until the given time.
+// If the request names a specific shard, this function will return an error
+// if that shard is already leased. Otherwise, this function will return the
+// index of the oldest shard for the given issuer.
+func (ssa *SQLStorageAuthority) LeaseCRLShard(ctx context.Context, req *sapb.LeaseCRLShardRequest) (*sapb.LeaseCRLShardResponse, error) {
+	if core.IsAnyNilOrZero(req.Until, req.IssuerNameID) {
+		return nil, errIncompleteRequest
+	}
+	if req.Until.AsTime().Before(ssa.clk.Now()) {
+		return nil, fmt.Errorf("lease timestamp must be in the future, got %q", req.Until.AsTime())
+	}
+
+	if req.MinShardIdx == req.MaxShardIdx {
+		return ssa.leaseSpecificCRLShard(ctx, req)
+	} else {
+		return ssa.leaseOldestCRLShard(ctx, req)
+	}
+}
+
+// leaseOldestCRLShard finds and then leases the crl shard for the given issuer
+// which is not currently leased and hasn't been updated in the longest. It
+// returns an error if all shards for that issuer are already leased.
+func (ssa *SQLStorageAuthority) leaseOldestCRLShard(ctx context.Context, req *sapb.LeaseCRLShardRequest) (*sapb.LeaseCRLShardResponse, error) {
+	shardIdx, err := db.WithTransaction(ctx, ssa.dbMap, func(txWithCtx db.Executor) (interface{}, error) {
+		var shards []*crlShardModel
+		_, err := txWithCtx.Select(
+			&shards,
+			`SELECT id, issuerID, idx, thisUpdate, nextUpdate, leasedUntil
+			FROM crlShards
+			WHERE issuerID = ?`,
+			req.IssuerNameID,
+		)
+		if err != nil {
+			return -1, err
+		}
+
+		// Convert the slice to a map, so we can easily detect shards that we expect
+		// to exist, but don't. At the same time, keep track of the oldest shard
+		// we've seen, so we can work on that one if none are missing.
+		shardMap := make(map[int]*crlShardModel, len(shards))
+		var oldest *crlShardModel
+		for _, shard := range shards {
+			if shard.Idx < int(req.MinShardIdx) || shard.Idx > int(req.MaxShardIdx) {
+				continue
+			}
+			shardMap[shard.Idx] = shard
+			if shard.LeasedUntil.After(ssa.clk.Now()) {
+				continue
+			}
+			if oldest == nil ||
+				(shard.ThisUpdate == nil && oldest.ThisUpdate != nil) ||
+				shard.ThisUpdate.Before(*oldest.ThisUpdate) {
```

Comment:
`oldest` is already a `*crlShardModel`. I think you could remove the pointer here on `*oldest`. Something seems weird w.r.t oldest/*oldest on the line above.

Reply:
This isn't dereferencing `oldest`, it's dereferencing `oldest.ThisUpdate`, which is a `*time.Time`. You're totally right that the syntax is confusing, but this is just the default order-of-operations across most C-like languages: we apply `.` before we apply `*`. The way a Programming Languages person would say it is that "the dot (.) operator _binds tighter_ than the dereference (*) operator".